drivers: regulator: add regulator shell

danieldegrasse opened the pull request:
Add a regulator shell. This shell supports enabling and disabling regulators, and also supports the extended PMIC regulator API when a PMIC device is present.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

henrikbrixandersen reviewed (changes requested):
Why not add a generic PMIC shell (or better yet, regulator shell) under `drivers/regulator/` that users can enable where needed?

danieldegrasse committed:
drivers: regulator: add regulator shell

Add shell for interacting with regulator devices, to enable
users to test power management settings quickly and verify that their
regulator is functioning as expected. Also allows users to interact with
regulator PMIC devices

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
PR updated to move the PMIC shell to a shell available within the regulators driver folder. The shell still only supports PMIC regulators, because there are parts of the PMIC API that are not present on standard regulators. The sample has been changed to simply enable this shell.

danieldegrasse commented:
@henrikbrixandersen Could you take another look at this?

henrikbrixandersen commented on drivers/regulator/Kconfig.pmic:
```diff
@@ -9,12 +9,22 @@ config REGULATOR_PMIC
+
+config REGULATOR_PMIC_SHELL
+	bool "PMIC regulator shell"
+	select SHELL
```
I'd use `depends on SHELL` along with `default y` here.

danieldegrasse replied:
Fixed in latest push

henrikbrixandersen commented on drivers/regulator/pmic_shell.c:
```diff
@@ -0,0 +1,169 @@
+
+LOG_MODULE_REGISTER(regulator_shell, CONFIG_REGULATOR_LOG_LEVEL);
+
+#define GET_REGULATOR(node) DEVICE_DT_GET(node),
```
What's the point here? You're just making an alias for an established, well-known macro?

danieldegrasse replied:
yeah, this is wrong- I've removed it

henrikbrixandersen commented on drivers/regulator/pmic_shell.c:
```diff
@@ -0,0 +1,169 @@
+
+#define GET_REGULATOR(node) DEVICE_DT_GET(node),
+
+static const struct device *supported_pmics[] = {
+	DT_FOREACH_STATUS_OKAY(regulator_pmic, GET_REGULATOR)
+	};
+static const struct device *pmic_dev;
```
Why not use a `<device>` argument to all `regulator` shell commands as done in other shells? There's no need to make this array at compile-time. Technically, other regulator drivers can than PMIC devices can be controlled via the exact same regulator API calls.

danieldegrasse replied:
Changed the commands to use a `device` argument directly

henrikbrixandersen commented on drivers/regulator/pmic_shell.c:
```diff
@@ -0,0 +1,169 @@
+	};
+static const struct device *pmic_dev;
+
+#define MAX_REGULATOR_VOLTAGE 5000000 /* Max voltage of 5V */
```
Why?

danieldegrasse replied:
This is legacy, since this shell was initially used to test the PCA9420, a PMIC regulator that won't go above that voltage. I've made the max voltage an argument to the shell command to set the PMIC voltage

henrikbrixandersen commented on drivers/regulator/pmic_shell.c:
```diff
@@ -0,0 +1,169 @@
+	struct onoff_client cli;
+
+	if (pmic_dev == NULL) {
+		shell_print(sh, "No PMIC device has been selected");
```
You need to use `shell_error()` for error printouts.

danieldegrasse replied:
Thanks, I've update this accordingly

henrikbrixandersen commented on drivers/regulator/pmic_shell.c:
```diff
@@ -0,0 +1,169 @@
+	shell_print(sh, "Enabling PMIC");
+	if (regulator_enable(pmic_dev, &cli) < 0) {
+		shell_print(sh, "Failed  to enable PMIC");
+		return -EACCES;
```
Why `-EACCESS`? Why not propagate the error returned (and mention it in the error output for easier debugging)?

danieldegrasse replied:
Good point- just tried to choose a logical error here. I've changed this to propagate the error

henrikbrixandersen commented on drivers/regulator/pmic_shell.c:
```diff
@@ -0,0 +1,169 @@
+	return 0;
+}
+
+static int cmd_list_devices(const struct shell *sh, size_t argc, char **argv)
+{
+	for (int i = 0; i < ARRAY_SIZE(supported_pmics); i++) {
+		shell_print(sh, "%s", supported_pmics[i]->name);
+	}
+	return 0;
+}
```
This is what `device list` is for.

henrikbrixandersen commented on drivers/regulator/pmic_shell.c:
```diff
@@ -0,0 +1,169 @@
+		cmd_list_devices),
+	SHELL_SUBCMD_SET_END
+);
+SHELL_CMD_REGISTER(pmic, &pmic_set, "PMIC Management", NULL);
```
I would prefer a generic top-level `regulator` command with `enable` and `disable` commands - and a conditional `pmic` subcommand for the PMIC specific commands. This will make the shell module useful for non-PMIC regulators as well.

danieldegrasse replied:
I've updated the shell to use this approach, so that generic regulators will work

henrikbrixandersen commented on samples/drivers/pmic_shell/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(pmic_shell)
```
We usually don't add a sample just for a shell module as it can easily be enabled where needed.

danieldegrasse replied:
Removed sample in latest push

carlescufi reviewed (dismissed):
LGTM

henrikbrixandersen commented on drivers/regulator/Kconfig:
```diff
@@ -8,6 +8,13 @@ menuconfig REGULATOR
 
+config REGULATOR_SHELL
+	bool "Regulator shell"
+	depends on SHELL
```
```suggestion
	default y
	depends on SHELL
```

danieldegrasse replied:
Fixed in latest push

henrikbrixandersen commented on drivers/regulator/Kconfig.pmic:
```diff
@@ -9,12 +9,15 @@ config REGULATOR_PMIC
 	  Enable the pmic regulator driver
 
+if REGULATOR_PMIC
+
 config PMIC_REGULATOR_INIT_PRIORITY
 	int "Regulator Init priority"
-	depends on REGULATOR_PMIC
 	default 75
 	help
 	  I2C based regulator init priority.
 	  Must be greater than KERNEL_INIT_PRIORITY_DEVICE so I2C is initialized,
 	  and less than SDMMC_INIT_PRIORITY, since the SDMMC driver uses an I2C
 	  PMIC regulator
+
+endif # REGULATOR_PMIC
```
`depends on REGULATOR_PMIC` is preferred when only a single `config` is affected.

danieldegrasse replied:
Updated to use `depends on REGULATOR_PMIC`

henrikbrixandersen commented on drivers/regulator/regulator_shell.c:
```diff
@@ -0,0 +1,154 @@
+	int lvol, uvol, ret;
+	const struct device *reg_dev;
+
+	if (argc != 4) {
+		shell_print(sh, "usage: %s "
+			"[pmic dev] [lower limit (mV)] [upper limit (mV)]",
+			argv[0]);
+		return -EINVAL;
+	}
```
Use `SHELL_CMD_ARG` for automatic argc checks.

danieldegrasse replied:
I've updated all shell commands in this PR to use `SHELL_CMD_ARG`, thanks for the pointer.

henrikbrixandersen commented on drivers/regulator/regulator_shell.c:
```diff
@@ -0,0 +1,154 @@
+	int lcur, ucur, ret;
+	const struct device *reg_dev;
+
+	if (argc != 4) {
+		shell_print(sh, "usage: %s "
+			"[pmic dev] [lower limit (uA)] [upper limit (uA)]",
+			argv[0]);
+		return -EINVAL;
+	}
```
Same here.

henrikbrixandersen commented on drivers/regulator/regulator_shell.c:
```diff
@@ -0,0 +1,154 @@
+	const struct device *reg_dev;
+	int ret;
+
+	if (argc != 2) {
+		shell_print(sh, "usage: %s [dev name]", argv[0]);
+		return -EINVAL;
+	}
```
Same here.

henrikbrixandersen commented on drivers/regulator/regulator_shell.c:
```diff
@@ -0,0 +1,154 @@
+	const struct device *reg_dev;
+	int ret;
+
+	if (argc != 2) {
+		shell_print(sh, "usage: %s [dev name]", argv[0]);
+		return -EINVAL;
+	}
```
Same here.

henrikbrixandersen commented on drivers/regulator/regulator_shell.c:
```diff
@@ -0,0 +1,154 @@
+		return -EINVAL;
+	}
+	reg_dev = device_get_binding(argv[1]);
+	if (reg_dev == NULL) {
```
You need to use `device_is_ready()` instead of checking against `NULL` all over.

henrikbrixandersen commented on drivers/regulator/regulator_pmic.c:
```diff
@@ -273,15 +273,13 @@ static int disable_regulator(const struct device *dev)
-		return onoff_sync_finalize(&data->srv, key, NULL, rc, true);
-	}
-	return onoff_sync_finalize(&data->srv, key, NULL, rc, true);
+	return onoff_sync_finalize(&data->srv, key, NULL, rc, false);
```
This is a bug fix and really ought to go into a separate PR.

danieldegrasse replied:
Done, PR number is #51092
